File: path-sum-ii.cpp
// Runtime: 12 ms
class Solution {
public:

    void solve(vector<vector<int> > &res, vector<int> temp, int tsum, TreeNode* root, int sum)
    {
        if(!root)
            return;

        temp.push_back(root->val);
        tsum += root->val;

        if(!root->left && !root->right && tsum == sum)
            res.push_back(temp);

        solve(res, temp, tsum, root->left, sum);
        solve(res, temp, tsum, root->right, sum);
    }
    vector<vector<int> > pathSum(TreeNode* root, int sum) {
        vector<vector<int> > res;
        vector<int> temp;
        solve(res, temp, 0, root, sum);
        return res;
    }
};

File: path-sum-ii.py
# Runtime: 78 ms
class Solution(object):

    def solve(self, root, sum, temp, res, tar):
        if(root):
            sum += root.val
            if not root.left and not root.right and tar == sum:
                res.append(temp + [root.val])

            self.solve(root.left, sum, temp + [root.val], res, tar)
            self.solve(root.right, sum, temp + [root.val], res, tar)

    def pathSum(self, root, sum):
        res = []
        temp = []
        self.solve(root, 0, temp, res, sum)
        return res
